datafusion/core/src/physical_plan/sorts/stream.rs

```diff
@@ -207,3 +233,237 @@ impl<T: FieldArray> PartitionedStream for FieldCursorStream<T> {
+            batches: HashMap::with_hasher(RandomState::new()),
+            streams,
+            reservation,
+        }
+    }
+
+    pub fn get_batches(&self, batch_ids: &[Uuid]) -> Vec<Arc<RecordBatch>> {
+        batch_ids.iter().map(|id| self.batches[id].clone()).collect()
+    }
+
+    pub fn remove_batches(&mut self, batch_ids: &[Uuid]) {
+        for id in batch_ids {
+            self.batches.remove(id);
+        }
+    }
+
+    fn poll_next(
+        &mut self,
+        cx: &mut Context<'_>,
+        stream_idx: usize,
+    ) -> Poll<Option<Result<(C, Uuid, BatchOffset)>>> {
+        Poll::Ready(ready!(self.streams.poll_next(cx, stream_idx)).map(|r| {
+            r.and_then(|(cursor, batch)| {
+                self.reservation.try_grow(batch.get_array_memory_size())?;
+                let batch_id = Uuid::new_v4();
+                self.batches.insert(batch_id, Arc::new(batch));
+                Ok((cursor, batch_id, BatchOffset(0_usize)))
+            })
+        }))
+    }
+}
+
+/// A newtype wrapper around a set of fused [`MergeStream`]
+/// that implements debug, and skips over empty inner poll results
+struct FusedMergeStreams<C>(Vec<Fuse<MergeStream<C>>>);
+
+impl<C: Cursor> FusedMergeStreams<C> {
+    fn poll_next(
+        &mut self,
+        cx: &mut Context<'_>,
+        stream_idx: usize,
+    ) -> Poll<Option<Result<(Vec<(C, Uuid, BatchOffset)>, Vec<SortOrder>)>>>
+    {
+        loop {
+            match ready!(self.0[stream_idx].poll_next_unpin(cx)) {
+                Some(Ok((_, sort_order))) if sort_order.len() == 0 => continue,
+                r => return Poll::Ready(r),
+            }
+        }
+    }
+}
+
+pub struct YieldedCursorStream<C: Cursor> {
+    // inner polled batch cursors, per stream_idx, which are partially yielded
+    cursors: Vec<Option<VecDeque<(C, Uuid, BatchOffset)>>>,
+    /// Streams being polled
+    streams: FusedMergeStreams<C>,
+}
+
+impl<C: Cursor + std::marker::Send> YieldedCursorStream<C> {
+    pub fn new(streams: Vec<MergeStream<C>>) -> Self {
+        let stream_cnt = streams.len();
+        Self {
+            cursors: (0..stream_cnt).map(|_| None).collect(),
+            streams: FusedMergeStreams(streams.into_iter().map(|s| s.fuse()).collect()),
+        }
+    }
+
+    fn incr_next_batch(
+        &mut self,
+        stream_idx: usize,
+    ) -> Option<(C, Uuid, BatchOffset)> {
+        self.cursors[stream_idx]
+            .as_mut()
+            .map(|queue| queue.pop_front())
+            .flatten()
+    }
+
+    // TODO: in order to handle sort_order, we need to either:
+    // parse further
+    // or concat the cursors
+    fn try_parse_batches(
+        &mut self,
+        stream_idx: usize,
+        cursors: Vec<(C, Uuid, BatchOffset)>,
+        sort_order: Vec<SortOrder>,
+    ) -> Result<()> {
+        let mut cursors_per_batch: HashMap<(Uuid, BatchOffset), C, RandomState> =
+            HashMap::with_capacity_and_hasher(cursors.len(), RandomState::new());
```

Comment:
Updated the hasher used, based upon the CPU profiler ran on release build.

BEFORE:
```
5.05 Gc == 100% at root of SortPreservingCascadeStream
24.8% is from `core::hash::BuildHasher`
```

AFTER:
```
4.53Gc == 100%
19.7% from hasher
```

Made the same change in the record batch tracker -- but that was not as significant.

Reply:
Moving from `UUid` to increasing `u64` id might bring a similar gain (see other comment).

Reply:
It did. Down to 3.8Gc on the same test case. TY!